File: kits/julia/set_kit.jl
"""
    set_kit([project_name, project_path])

Create a folder named 'project_name' in 'project_path' and set a file structure 
and create a Pluto notebook inside it.

File structure and notebooks are detailed [here](https://github.com/ludmillafigueiredo/computational_notebooks).

# Example
```julia-repl
 julia> include("set_kit.jl")
 julia> set_kit.jl
```
"""
function set_kit(;project_name::String = "project", project_path::String = pwd())
    ## Create project in path
    cd(project_path)
    ## Create file structure
    ### first level
    mkdir(project_name)
    #### TODO: add README per folder
    ### main folders of a project
    map(mkdir, map(x -> joinpath(project_name, x),
                   ["results","text", "submission"]))
    ### main folders of results
    map(mkdir, map(x -> joinpath(joinpath(project_name,"results"), x),
                   ["raw_data","clean_data", "scripts", "semi_products"]))
    ### main folders of text
    map(mkdir, map(x -> joinpath(joinpath(project_name,"text"), x),
                   ["figures","tables", "supplementary"]))

    ## Create the README files
    cp("../READMEs/main.md", joinpath(project_name, "README.md"), force = false)
    cp("../READMEs/results.md", joinpath(project_name,"results","README.md"), force = false)
    cp("../READMEs/text.md", joinpath(project_name, "text", "README.md"), force = false)
    cp("../READMEs/submission.md", joinpath(project_name, "submission", "README.md"), force = false)
    
    ## Create the minimal notebook
    touch(joinpath(project_name, "notebook.jl"))
    open(joinpath(project_name, "notebook.jl"), "w") do io
        write(io, "### A Pluto.jl notebook ###")
        write(io, "\n\n")
        write(io, "using Markdown")
        write(io, "\n")
        write(io, "using InteractiveUtils")
        write(io, "\n\n")
        write(io, "# ╔═╡ 307335ba-cf46-11eb-28b0-f199cf048ae6")
        write(io, "\n")
        write(io, "md\" ## Your title here\"")
        write(io, "\n\n")
        write(io, "# ╔═╡ 37c7bffc-cf46-11eb-19b8-4175e38e818b")
        write(io, "\n")      
        write(io, "## Paths for easy call from the Pluto notebook")
        write(io, "\n")
        write(io, "begin")
        write(io, "\n")
        write(io, "\tdata_dir = joinpath(\"results\", \"clean_data\") ## do NOT play with stuff in raw_data. That is your back-up")
        write(io, "\n")
        write(io, "\tscripts_dir = joinpath(\"results\", \"scripts\")")
        write(io, "\n")
        write(io, "\tsuppl_dir = joinpath(\"results\", \"supplementary\")")
        write(io, "\n")
        write(io, "\tsemiprods_dir = joinpath(\"results\", \"semi_products\")")
        write(io, "\n")
        write(io, "\tfigures_dir = joinpath(\"text\", \"figures\")")
        write(io, "\n")
        write(io, "\ttables_dir = joinpath(\"text\", \"tables\")")
        write(io, "\n")
        write(io, "end")
        write(io, "\n\n\n")
        write(io, "# ╔═╡ Cell order:")
        write(io, "\n")
        write(io, "# ╠═307335ba-cf46-11eb-28b0-f199cf048ae6")
        write(io, "\n")
        write(io, "# ╠═37c7bffc-cf46-11eb-19b8-4175e38e818b")
    end
end

File: kits/R-Python/set_kit.R
#' Set a minimal file structure and create the notebook
#'
#' @param lang a string, the language \"r\" or \"python\", you want to create your notebook in.  
#'  It defaults to "r".
#' @param project_name a string, the name the folder of the project. 
#'  It defaults to \"project\".
#' @param project_path a string, the path where the project folder should be created. 
#'  It defaults to the current current working directory.
#'  
#' @details The file structure and notebooks are detailed here \url{https://github.com/ludmillafigueiredo/computational_notebooks}.
#' 
#' @example
#' source("set_kit.R")
set_kit <- function(lang = "r", project_name = "project", project_path = getwd()){
  
  ## Create project in path 
  dir.create(file.path(project_path, project_name), recursive = TRUE)
  
  ## Set working directory
  setwd(file.path(project_path, project_name))
  
  ## Create file structure
  ### first level
  #### TODO: add README per folder
  ### main folders of a project
  sapply(c("results","text", "submission"), dir.create)
  ### main folders of results
  sapply(file.path(paste(file.path("results"), 
                         c("data", "scripts"),
                         sep = "/")), 
         dir.create)
  sapply(file.path(paste(file.path("results", "data"), 
                         c("raw", "processed"),
                         sep = "/")), 
         dir.create)
  ### main folders of text
  sapply(file.path(paste(file.path("text"), 
                         c("figures","tables", "supplementary"),
                         sep = "/")), 
         dir.create)
  
  ## Create the README files
  file.copy(from = "../READMEs/main.md", to = "README.md", overwrite = FALSE)
  file.copy(from = "../READMEs/results.md", to = "results/README.md", overwrite = FALSE)
  file.copy(from = "../READMEs/text.md", to = "text/README.md", overwrite = FALSE)
  file.copy(from = "../READMEs/submission.md", to = "submission/README.md", overwrite = FALSE)
  
  ## Create the minimal notebook
  sink("notebook.Rmd")
  cat("---")
  cat("\n")
  cat("title: \"Your title here\"")
  cat("\n")
  cat("output: pdf_document")
  cat("\n")
  cat("---")
  cat("\n")
  cat("<!-- ---------------------------------- VERY MINIMAL INTRO TO RNOTEBOOKS ---------------------------------- -->")
  cat("\n\n")
  cat("<!-- Just erase this whole block, if you already know how it works) -->")
  cat("\n\n")
  cat("<!-- The basic idea of notebooks is having your text written without any marking: -->")
  cat("\n\n")
  cat("\"Uncommented\" text to show you how it works.")
  cat("\n\n")
  cat("<!-- and code is put inside blocks like this (also called 'chunks'): -->")
  cat("\n\n")
  cat(paste("```{", lang, "}", sep = ""))
  cat("\n")
  cat("print(\"This is an example of a code block\")")
  cat("\n")
  cat("```")
  cat("\n")
  cat("<!-- Very important, check the options of how much of your code and outputs you want to show when compiling your notebook: -->")
  cat("<!-- https://bookdown.org/yihui/rmarkdown/r-code.html -->")
  cat("\n\n")
  cat("<!-- As explained above, these are defined in your code chunks. -->")
  cat("\n\n")
  cat("<!-- ----------------------------- END OF VERY MINIMAL INTRO TO RNOTEBOOKS -------------------------------- -->")
  cat("\n\n")
  cat("<!-- Create objects for easy access to the folders related to the project. -->")
  cat("\n")
  cat("<!-- The code in this chunk is not relevant for the reader, and thus is not included in the knitted version. -->")
  cat("\n")
  cat("<!-- This is also useful if you are not using the folders suggested by this kit, and want to preserve your privacy. -->")
  cat("\n")
  cat(paste("```{", lang, " set-up, include = FALSE}", sep = ""))
  cat("\n")
  cat("data_dir <- file.path(\"results\", \"clean_data\") ## do NOT play with stuff in raw_data. That is your back-up")
  cat("\n")
  cat("scripts_dir <- file.path(\"results\", \"scripts\")")
  cat("\n")
  cat("suppl_dir <- file.path(\"results\", \"supplementary\")")
  cat("\n")
  cat("semiprods_dir <- file.path(\"results\", \"semi_products\")")
  cat("\n")
  cat("figures_dir <- file.path(\"text\", \"figures\")")
  cat("\n")
  cat("tables_dir <- file.path(\"text\", \"tables\")")
  cat("\n")
  cat("```")
  cat("\n\n")
  cat("# Read your data")
  cat("\n")
  cat(paste("```{", lang, "}", sep = ""))
  cat("\n")
  cat("## data <- read.csv(file.path(data_dir)")
  cat("\n")
  cat("```")
  cat("\n\n")
  cat("# Explore it")
  cat("\n\n")
  cat("# Make your main figures/tables")
  cat("\n")
  cat(paste("```{", lang, "}", sep = ""))
  cat("\n")
  cat("## ggsave(file.path(data_dir, \"plot1.png\")")
  cat("\n")
  cat("```")
  cat("\n\n")
  cat("*R version, the OS and attached or loaded packages:*")
  cat("\n")
  cat("<!-- Leave this so people know the software they need to reproduce your work. -->")
  cat("\n")
  cat("```{r}")
  cat("\n")
  cat("sessionInfo()")
  cat("\n")
  cat("```")
  sink()
}
